### PR TITLE
Optional suspend route pruning if NATS unavailable

### PR DESCRIPTION
**Pull request by markstgodard**
Currently, in the event that a NATS cluster goes down or becomes unavailable, gorouter will continue to prune routes. More importantly, gorouter will actually attempt to reconnect to all configured NATS servers and if exceeds the NATS client MaxReconnect attempts, it will exit and monit will keep restarting gorouter. In this catastrophic scenario, gorouter will not retain any routing info and apps will not be routable.

This approach favors consistency over availability, however if an operator would prefer to keep apps routable in the event of a NATS cluster outage, this PR provides that option.

The PR does the following:
- Opt-in config **suspend_pruning_if_nats_unavailable** to suspend route pruning if we cannot connect to NATS
- Set max reconnect in NATS client to -1 (no limit) when suspend prune enabled
- if pruning was suspended at any point and when NATS becomes available, the routes 'updated' timestamp will be marked with a current timestamp, so that it "resets the clock" for routes and avoids gorouter from pruning all the routes once NATS comes back online
- Default behavior will remain as pruning routes and not override max reconnect (current behavior)

_NOTE: This strategy favors availability over consistency and there is a possibility of routing to an incorrect endpoint in the case of port re-use. To be used with caution._

Once this PR is merged, I can update cf-release spec, templates to include **router.suspend_pruning_if_nats_unavailable** with the default value being **false**.

**Acceptance:**
- enable "suspend_pruning_if_nats_unavailable" setting to true (can change by hand in gorouter.yml or via manifest change)
- restart gorouter
- start watching the gorouter routing table
i.e.
```
while true; do clear && curl -s http://router:router@IP_ADDRESS_OF_GOROUTER:8080/routes | jq . && sleep 3; done
```
- bring down all NATS instances one by one
- gorouter.log should indicate "prune-suspended"
- wait for period of time until prune stale droplet threshold exceeded (or longer)
- routes should still be showing up in watch of routing table
- bring back nats instances
- routes should remain showing up in watch of routing table

### CODE REVIEW

**Comment by shalako** [comment]
Hi Mark,
Just back from vacation and reviewing this. 

- Set max reconnect in NATS client to -1 (no limit) when suspend prune enabled

what does max reconnect mean?

Thanks

**Comment by markstgodard** [comment]
Max reconnect set to -1 in nats client means to always try reconnecting. Our current setup will abort after 10 attempts and then our callback handler in gorouter will exit. Then monit does its thing with gorouter process

**Comment by shalako** [comment]
Do you mean that after 10 attempts gorouter effectively crashes and when monit restarts it, gorouter will try 10 more times? 

If this new property is set to true, and NATS is unavailable on a deploy, will the deploy fail if the routers are restarted? We like to give operators fast feedback while they are in the BOSH context. 

Maybe we should do the continuous retry all the time, rather than only when this property is set to true. But on a deploy, where the routers are bounced, if NATS is unavailable then the deploy should fail as a result of router failing to come up.

Is it possible to fail the deploy but leave the router in a running state, continuously attempting to contact NATS? I'm guessing in order to fail the deploy we have to report the process failing to monit.

**Comment by markstgodard** [comment]
Yes, once we exceed the default nats MaxReconnect, we receive a Closed callback which gorouter has registered a func to log and exit. Monit then sees it as failed and does the restart dance. 

With MaxReconnect set to -1 (in the case we want to suspend pruning for nats) the connect to Nats logic on gorouter behaves the same way, we will fail. We have special logic on gorouter startup to try 3 times and if we did not connect successfully then we exit and monit does its thing.  

The MaxReconnect  option set to -1 comes into play when we have made at least a successful connect once at startup. 

So the fail fast behavior on startup is still in place. 

Cheers

**Comment by shalako** [comment]
If `suspend_pruning_if_nats_unavailable:true`, then `MaxReconnect: -1`. In one comment above you say that this results in gorouter will continue retrying to connect to NATS forever. Then you say this config results in gorouter will fail, triggering monit to restart. I'm confused.

To confirm; the startup behavior is the same, no matter what `suspend_pruning_if_nats_unavailable` is set to?

**Comment by markstgodard** [comment]
Correct. Startup behavior is the same no matter what suspend_pruning_if_nats_is_unavailable is set to. After successful startup, then the max reconnect -1 will have meaning and will not fail and exit after exceeds the default 10 retries. Sorry should have clarified that startup had same semantics 
Cheers

**Comment by shalako** [comment]
> If suspend_pruning_if_nats_unavailable:true, then MaxReconnect: -1. In one comment above you say that this results in gorouter will continue retrying to connect to NATS forever. Then you say this config results in gorouter will fail, triggering monit to restart. I'm confused.

>  After successful startup, then the max reconnect -1 will have meaning and will not fail and exit after exceeds the default 10 retries.

Please clarify whether there is a difference in behavior with regard to gorouter exiting and being restarted by monit when `suspend_pruning_if_nats_is_unavailable` is true of false. Does gorouter exit in one case but not in another?

**Comment by markstgodard** [comment]
Hi @shalako 

The `suspend_pruning_if_nats_is_unavailable` setting will NOT affect the gorouter initial startup behavior related to failing fast if gorouter cannot connect to NATS. 

On initial startup, if gorouter cannot connect to NATS server(s), gorouter will exit with a non-zero exit code  (regardless of the `suspend_pruning_if_nats_is_unavailable` setting **true** or **false**)

The `suspend_pruning_if_nats_is_unavailable` setting will only come into play once gorouter has started up successfully (meaning it has connected to NATS on startup).

If the gorouter has successfully started up and if all NATS server(s) go away for an extended period of time and if `suspend_pruning_if_nats_is_unavailable` = **true**, then gorouter will not prune routes  and will continue attempting to connect to NATS servers (nats client max reconnect -1 )

If the gorouter has successfully started up and if all NATS server(s) go away for an extended period of time and if `suspend_pruning_if_nats_is_unavailable` = **false**, then gorouter will exit when the nats client connection closes after exceeds max reconnects (nats client max reconnect set to nats client default [60](https://github.com/nats-io/nats/blob/master/nats.go#L34))


**TL;DR;**
- gorouter startup behavior of "fail fast" is the same
- gorouter will eventually exit if `suspend_pruning_if_nats_is_unavailable` = **false** and if NATS goes away for an extended period of time
- gorouter will NOT exit if `suspend_pruning_if_nats_is_unavailable` = **true** and if NATS goes away for an extended period of time

Feel free to ping me if you wanted to clarify anything

Cheers

**Comment by shalako** [comment]
Thank you for the clarification. Prioritizing for teams' review.

File: registry/registry_test.go [hunk 1]
```diff
@@ -687,3 +736,13 @@ var _ = Describe("RouteRegistry", func() {
 		Expect(string(marshalled)).To(Equal(`{}`))
 	})
 })
+
+func portListening(addr string) bool {
```
Comment (crhino) [inline registry/registry_test.go]: @markstgodard Seems like these function is not used anywhere?

Comment (markstgodard) [inline registry/registry_test.go]: Hey @crhino, good catch... it was a test helper function, I must have refactored the test and forgot to dd it.  I can nix it.

**Comment by crhino** [comment]
@shalako Are you OK with the name `suspend_pruning_if_nats_is_unavailable`? We are thinking about the possibility of implementing the same feature for Routing API, and having to add another property like `suspend_pruning_if_routing_api_is_unavailable`.

@markstgodard Commented on the PR itself, seems like an unused test function snuck in, but other than that everything looks good.

**Comment by markstgodard** [comment]
@shalako @crhino We could have a more generic name for the property. And then in subsequent work, could add routing-api as another source.

**Comment by shalako** [comment]
I'm ambivalent about the name. I like that the current name is descriptive. If we used a generic name (also ok) the description would have to note that if the routing api is unavailable and NATS is available, routes will be pruned.

**Comment by crhino** [comment]
@flawedmatrix and I discussed and decided we also like the descriptive nature of the current name. If we need to add another property later for routing-api that is not so bad.

**Comment by jzlwang** [comment]
@markstgodard you mentioned that when `suspend_pruning_if_nats_is_unavailable` is true, we should see a "prune-suspended" message in `gorouter.log`. While doing acceptance I didn't see any logs at all in `gorouter.log` or `gorouter.err.log`. Can you please fix this for us? Thanks.

**Comment by markstgodard** [comment]
@jzlwang Hey Judy, right I had that log statement as `debug` level, so the default logging level would not show that.  Did we what that log statement as `info` level?  Seems reasonable.

**Comment by markstgodard** [comment]
Hey Judy, I don't push access anymore, so I had to submit a PR:
https://github.com/cloudfoundry/gorouter/pull/145
Also I didnt want to add commits to this specific PR since its closed.

Just get one of the team to review and merge if you want this to show up as info level.

Cheers

**Comment by jzlwang** [comment]
Awesome, thanks Mark!

**Comment by CAFxX** [comment]
@flawedmatrix @markstgodard I haven't tried yet the fixed version but from a check of this fix I'd say it won't solve the problem I reported in #132 (that was closed because #140 was merged)
